appcelerator/titanium_mobile
# refactor(android): Ti.Filesystem external storage handling

**JIRA:**
- https://jira.appcelerator.org/browse/TIMOB-28231
- https://jira.appcelerator.org/browse/TIMOB-28230
- https://jira.appcelerator.org/browse/TIMOB-28182

**Summary:**
- [TIMOB-28231](https://jira.appcelerator.org/browse/TIMOB-28231) Changed `Ti.Filesystem.externalStorageDirectory` to use scoped storage.
  * Only directory location is not accessible when targeting and running on Android 11.
  * Changed to a new location which does not require `WRITE_EXTERNAL_STORAGE` permission.
- [TIMOB-28230](https://jira.appcelerator.org/browse/TIMOB-28230) Added property `Ti.Filesystem.externalCacheDirectory`.
  * Intended to store large temporary files.
- [TIMOB-28182](https://jira.appcelerator.org/browse/TIMOB-28182) Modified build system to only add `WRITE_EXTERNAL_STORAGE` permission when needed.
  * Only adds permission if using API: `Ti.Fileystem.requestStoragePermission()`, `Ti.Media.showCamera()`, `Ti.Media.saveToPhotoGallery()`.
- Added new `TiMockFile` Java class for handling inaccessible/invalid file URLs.
  * Handles case where we can't fetch external storage path if volume is not mounted. (Natively returns null.)
  * This way `Ti.Filesystem.getFile()` no longer returns `null` and instead returns a file object which no-ops.

---
**API Test:**
1. Build and run `FileExternalTest.js` attached to [TIMOB-28231](https://jira.appcelerator.org/browse/TIMOB-28231) on Android.
2. Tap on the "Write" button.
3. Verify the below is shown on screen.
`External Storage File`
`Write Time: <CurrentTime>`
`External Cache File`
`Write Time: <CurrentTime>`
4. Tap on the "Back" button.
5. Re-launch the app.
6. Verify the above 4 lines of text are shown.
7. Tap on the "Back" button.
8. Go to: `Setting -> Apps -> App Info -> [AppName] -> Storage`
9. Tap on the `Clear Cacher` button.
10. Re-launch the app.
11. Verify only 2 lines are shown. The "External Cache File" text should **NOT** be shown.
`External Storage File`
`Write Time: <CurrentTime>`
12. Tap on the "Delete" button. _(Will delete external storage files.)_
13. Back out of the app.
14. Re-launch the app.
15. Verify no text is shown.

---
**General Permission Test:**
1. Build `FileExternalTest.js` attached to [TIMOB-28231](https://jira.appcelerator.org/browse/TIMOB-28231) for Android.
2. Open file under: `./build/android/app/intermediates/merged_manifests/debug/AndroidManifest.xml`
3. Verify the file does **NOT** contain the `WRITE_EXTERNAL_STORAGE` permission.
4. Add the below permission to your "tiapp.xml" file.
5. Build for Android.
6. Verify "AndroidManifest.xml" file does contain the added permission.

```xml
<ti:app>
	<android>
		<manifest>
			<uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE"/>
		</manifest>
	</android>
</ti:app>
```

---
**requestStoragePermissions() Test:**
1. Build and run `RequestStoragePermissionsTest.js` attached to [TIMOB-28182](https://jira.appcelerator.org/browse/TIMOB-28182) on Android.
2. Tap on the "Request Storage Permission" button.
3. Verify a permission request dialog is shown.
4. Tap on the "Allow" button.
5. Verify the onscreen "hasPermissions:" label is set `true`.
6. Open file under: `./build/android/app/intermediates/merged_manifests/debug/AndroidManifest.xml`
7. Verify the file contains the following line.
`<uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE"/>`

---
**saveToPhotoGallery() Test:**
1. Build and run `ImageToGalleryTest.js` attached to [TIMOB-28182](https://jira.appcelerator.org/browse/TIMOB-28182) on Android 10 or 11.
2. Tap on the "Save Screenshot to Gallery" button.
3. A permission request dialog will only appear on Android 9 and older.
4. Go to the Photos gallery app.
5. Verify a screenshot of the app is in the gallery.
6. Open file under: `./build/android/app/intermediates/merged_manifests/debug/AndroidManifest.xml`
7. Verify the file contains the following. _(Notice the `maxSdkVersion` part.)_
`<uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" android:maxSdkVersion="28"/>`
8. Repeat steps 1-5 on Android 9 or older.

---
**showCamera() Test:**
1. Build and run `CameraPhotoExternalTest.js` attached to [TIMOB-28182](https://jira.appcelerator.org/browse/TIMOB-28182) on Android 10 or 11.
2. Tap on the "Take Picture" button.
3. A permission request dialog will only appear on Android 9 and older.
4. Take a photo.
5. Verify the photo is shown in your app.
6. Repeat steps 1-5 on Android 9 or older.
